File: solutions/csharp/tracks-on-tracks-on-tracks/1/TracksOnTracksOnTracks.cs
using System.Dynamic;

public static class Languages
{
    public static List<string> NewList() => new List<string>();

    public static List<string> GetExistingLanguages() => ["C#", "Clojure", "Elm"];

    public static List<string> AddLanguage(List<string> languages, string language) =>
        languages.Concat([language]).ToList();

    public static int CountLanguages(List<string> languages) => languages.Count;

    public static bool HasLanguage(List<string> languages, string language) =>
        languages.Contains(language);

    public static List<string> ReverseList(List<string> languages) =>
        languages.AsEnumerable().Reverse().ToList();

    public static bool IsExciting(List<string> languages)
    {
        if (languages.Count >= 1 && languages.First() == "C#")
            return true;
        if (languages.Count is 2 or 3)
            return languages[0..2].Contains("C#");
        return false;        
    }

    public static List<string> RemoveLanguage(List<string> languages, string language) =>
        languages.Where(l => l != language).ToList();

    public static bool IsUnique(List<string> languages)
    {
        var set = new HashSet<string>(languages);
        return set.Count == languages.Count;
    }
}

File: solutions/fsharp/tracks-on-tracks-on-tracks/1/TracksOnTracksOnTracks.fs
module TracksOnTracksOnTracks

let newList: string list = []

let existingList: string list = ["F#"; "Clojure"; "Haskell"]

let addLanguage (language: string) (languages: string list): string list =
    language :: languages

let countLanguages (languages: string list): int = languages.Length

let rec reverseList(languages: string list): string list = 
    match languages with
    | [] -> []
    | [x] -> [x]
    | x :: xs -> reverseList xs @ [x]

let excitingList (languages: string list): bool = 
    match languages with 
    | [] -> false
    | x :: xs -> 
        if x = "F#" then true
        else 
            match xs with
            | [] -> false
            | y :: ys -> y = "F#" && ys.Length < 2
